### ISSUE TITLE
Missing support for some penalty metrics, beam search, and top-k in `LogitsProcessor`.

### ISSUE THREAD

**Issue by EricLBuehler** [huggingface/candle#1263]
Hello everybody,

I have been working on adding sampling support to `candle-vllm`, but have noticed there are some features that could be added to `LogitsProcessor`. Although techniques like `top-p`, `multinomial`, and `argmax` are implemented, I this that the following features would be great to support.

Namely:
1) Penalty metrics (`presence_penalty`, `frequency_penalty`, and `repetition_penalty`)
    - These are important hyperparameters when sampling logits.
2) Beam search
    - Beam search is a popular sampling method.
3) `top-k` sampling
    - This is another popular sampling method.

What do you think about adding these sampling features?

**Comment by EricLBuehler** [comment]
See candle-sampling, for future reference.
